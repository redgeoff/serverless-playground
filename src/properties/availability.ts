import { APIGatewayProxyEvent } from "aws-lambda";
import * as yup from 'yup';
import { successResponse, runWarm, badRequestResponse } from '../utils';

const schema = yup.object().shape({
  // TODO: This is just a proof of concept and not necessarily the proper format for the id
  propertyId: yup.number().required().integer()
});

export const isValid = async (pathParameters: APIGatewayProxyEvent['pathParameters']) => schema.isValid(pathParameters)

const availability = async (event: APIGatewayProxyEvent /* , context */) => {
  if (!(await isValid(event.pathParameters))) {
    // TODO: use yup to actually report details of why the data is invalid
    return badRequestResponse({ errorMessage: 'yup says no' })
  }

  const { propertyId=undefined } = event.pathParameters || {};
  if (propertyId === '123') {
    return badRequestResponse({ errorMessage: 'cannot be 123' })
  }
  return successResponse({ foo: 'bar', propertyId });
};

export function foo(thing: string) {
  return thing
}

// runWarm function handles pings from the scheduler so you don't
// have to put that boilerplate in your function.
export default runWarm(availability);
