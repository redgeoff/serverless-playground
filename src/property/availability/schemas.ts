export const propertyIdSchema = {
  $id: 'http://example.com/product.schema.json',
  $schema: 'http://json-schema.org/draft-07/schema#',
  // title: 'Property',
  // description: 'A property',
  properties: {
    propertyId: {
      // description: 'The unique identifier for a property',
      type: 'string',
    },
  },
  required: ['propertyId'],
  type: 'object',
};
