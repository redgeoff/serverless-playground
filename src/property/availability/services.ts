import ResponseError from '../../utils/response-error';

export interface Availability {
  foo: string;
  propertyId?: string;
}

export class BadPropertyError extends ResponseError {
  constructor(message: string) {
    super('BAD_PROPERTY', message);
  }
}

export type GetAvailability = (propertyId?: string) => Promise<Availability>;

// Note: async is not needed here, but is used as a lot of real-world examples will need async
export const availability: GetAvailability = async (propertyId?: string) => {
  if (propertyId === '123') {
    throw new BadPropertyError('cannot be 123');
  }

  return { foo: 'bar', propertyId };
};
