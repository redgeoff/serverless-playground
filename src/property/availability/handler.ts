import handler, { HandlerEvent } from '../../utils/handler';
import { propertyIdSchema } from './schemas';
import {
  availability as availabilityService,
  GetAvailability,
} from './services';

async function availabilityHandler(
  availability: GetAvailability,
  event: HandlerEvent /* , context */
) {
  const { propertyId } = event.pathParameters;
  return availability(propertyId);
}

export function availabilityHandlerFactory(availability: GetAvailability) {
  return (event: HandlerEvent /* , context */) =>
    availabilityHandler(availability, event);
}

export const factoryProperties = {
  groups: { internalApp: ['INTERNAL_GET_AVAILABILITY'] },
  handlerFactory: availabilityHandlerFactory(availabilityService),
  schemas: { pathParameters: propertyIdSchema },
};

export default handler.factory(factoryProperties);
