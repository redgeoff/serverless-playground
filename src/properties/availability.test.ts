import availability, { foo, isValid } from './availability';

describe('availability', () => {
  it('executes as expected', async () => {
    expect(await availability({ pathParameters: { propertyId: '1234' } }, {})).toMatchSnapshot();
  });

  it('is valid', async () => {
    expect(await isValid({ propertyId: 'aaa' })).toEqual(false);
    // TODO: ...
  });

  it('should foo', () => {
    // function yo(cool: string) {
    //   return cool;
    // }
    // expect(foo());
    expect('bar').toEqual('bar');
  });
});
