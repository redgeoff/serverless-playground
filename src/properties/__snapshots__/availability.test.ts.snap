// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`availability executes as expected 1`] = `
Object {
  "body": "{\\"foo\\":\\"bar\\",\\"propertyId\\":\\"1234\\"}",
  "statusCode": 200,
}
`;
